### ISSUE TITLE
aws_ami + local_exec forces new resource

### ISSUE THREAD

**Issue by rafaljanicki** [hashicorp/terraform-provider-aws#2769]
### Terraform Version
Terraform v0.11.1

### Affected Resource(s)
Please list the resources as a list, for example:
- aws_ami
- aws_instance

### Terraform Configuration Files
```hcl
resource "null_resource" "bastion_packer" {
  provisioner "local-exec" {
    command = "cd ../packer && packer build -var region=${var.region} bastion.json"
  }
}

data "aws_ami" "bastion_ami" {
  depends_on = ["null_resource.bastion_packer"]

  most_recent = true
  owners = ["self"]

  filter {
    name = "tag:Application"
    values = ["the-bastion"]
  }
}

resource "aws_instance" "bastion_host" {
  ami = "${var.bastion_ami_id}"
  instance_type = "t2.nano"
  vpc_security_group_ids = ["${aws_security_group.bastion_security_group.id}"]
  subnet_id = "${var.public_subnet}"
  key_name = "${module.bastion_key_pair.key_name}"

  tags {
    Application = "Bastion"
    Name = "Bastion-${terraform.workspace}"
    Environment = "${terraform.workspace}"
    Terraform = true
  }
}
```

Note that for simplicity reasons, I haven't pasted whole modules, but only the relevant snippets hence the differences between references (`var.bastion_ami_id` vs `data.aws_ami.bastion_ami.id`)

### Debug Output
https://gist.github.com/rafaljanicki/d26817f07eb3175db22469b430b9ebb6

### Expected Behavior
The AMI ID doesn't change between runs (even the local-exec command doesn't run), so Terraform should not replace the Bastion instance with a new one as the values don't differ

### Actual Behavior
Terraform replaces the current Bastion with a new one, but with the same AMI ID

I do have an idea why this happens - TF can't derive the AMI ID as the aws_ami data depends on an external command, but TF does know that the command should not be run (nothing has changed in the null_resource), so, therefore, TF should know that nothing would change further

### Steps to Reproduce
1. `terraform apply`

**Comment by jasonwalsh** [comment]
I know that this is a late reply, but I'm doing exactly what @rafaljanicki is doing. I solved this by adding a tag to my AMI with the `null_resource.bastion_packer.id` value.

`packer.json`

```json
...
      "tags": {
        "id": "{{ user `id` }}"
      }
...
```

`main.tf`

```hcl
data "aws_ami" "ami" {
  filter {
    name   = "tag:id"
    values = ["${null_resource.bastion_packer.id}"]
  }

  most_recent = true
  owners      = ["self"]
}

resource "null_resource" "bastion_packer" {
  provisioner "local-exec" {
    command = "cd ../packer && packer build -var region=${var.region} -var id=${self.id} bastion.json"
  }
}
```

The trick is to use the `self` attribute of the provisioner which tags the AMI with a unique identifier. Once the resource is created it creates an implicit dependency in the `aws_ami` data source object without having to use the `depends_on` directive.

Hope this helps.

**Comment by jholloway7** [comment]
@jasonwalsh - I think with your trick you still need to use `depends_on` in the data source -- at least the first time you apply -- or you'll get this error during the plan phase:

    Error: Your query returned no results. Please change your search criteria and try again.

But even with `depends_on`, this workaround doesn't appear to address the issue being reported. The `aws_instance` still gets flagged as needing replaced because of the "(known after apply)" on the `ami_id` even when it doesn't change between runs.

It seems like to make it work the `ami_id` has to be static before the apply.

**Comment by cmawhorter** [comment]
i think this may be resolved in v0.12. i found this post while researching how to use packer with tf and used OPs file as a reference to create my own. it seems to work and does not run packer every apply. to run packer, i taint the null resource and it runs.
